### PR TITLE
Handle changes to the logical children collection while operating on it

### LINKED ISSUE #12799: Avalonia 11.0.4  internal error after closing the main application window
Avalonia 11.0.4 internal error after closing the main application window.

Message: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')

StackTrace:   
at System.ThrowHelper.ThrowArgumentOutOfRange_IndexException()
   at Avalonia.Collections.AvaloniaList`1.get_Item(Int32 index)
   at Avalonia.StyledElement.OnDetachedFromLogicalTreeCore(LogicalTreeAttachmentEventArgs e)
   at Avalonia.StyledElement.OnDetachedFromLogicalTreeCore(LogicalTreeAttachmentEventArgs e)
   at Avalonia.StyledElement.OnDetachedFromLogicalTreeCore(LogicalTreeAttachmentEventArgs e)
   at Avalonia.Controls.TopLevel.HandleClosed()
   at Avalonia.Controls.WindowBase.HandleClosed()
   at Avalonia.Controls.Window.HandleClosed()
   at Avalonia.Win32.WindowImpl.AppWndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
   at Avalonia.Win32.WindowImpl.WndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)

- OS: [Windows, Linux]
- Version [11.0.4]

avalonia_error

### LINKED ISSUE #13497: `ArgumentOutOfRangeException` in `OnDetachedFromLogicalTreeCore`
`OnDetachedFromLogicalTreeCore` enumerates children of the current control, but doesn't account for the collection of children changing during enumeration.

## To Reproduce

1. Start Control Gallery
2. Switch to the TabControl tab
3. Enable "Set TabItem.ContentTemplate"
4. Close Control Gallery

```
System.ArgumentOutOfRangeException
  HResult=0x80131502
  Message=Index was out of range. Must be non-negative and less than the size of the collection. Arg_ParamName_Name
  Source=System.Private.CoreLib
  StackTrace:
   at System.ThrowHelper.ThrowArgumentOutOfRange_IndexException()
   at System.Collections.Generic.List`1.get_Item(Int32 index)
   at Avalonia.Collections.AvaloniaList`1.get_Item(Int32 index) in Avalonia\src\Avalonia.Base\Collections\AvaloniaList.cs:line 148
   at Avalonia.StyledElement.OnDetachedFromLogicalTreeCore(LogicalTreeAttachmentEventArgs e) in Avalonia\src\Avalonia.Base\StyledElement.cs:line 932
   at Avalonia.StyledElement.OnDetachedFromLogicalTreeCore(LogicalTreeAttachmentEventArgs e) in Avalonia\src\Avalonia.Base\StyledElement.cs:line 934
   ...
```

## Expected behavior
`OnDetachedFromLogicalTreeCore` should account for the possibility of its children changing during enumeration.

## Environment 

- OS: N/A
- Avalonia-Version: 11.0.5

### LINKED ISSUE #14437: `ArgumentOutOfRangeException` in `OnDetachedFromVisualTreeCore`
## Describe the bug

https://github.com/AvaloniaUI/Avalonia/blob/88967de49e083992e9ee748d7767d7a999e42f41/src/Avalonia.Base/Visual.cs#L554-L563

An `ArgumentOutOfRangeException` can be thrown in the iteration of `visualChildren` if the underlying list changes.

```
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.Collections.Generic.List`1.get_Item(Int32 index)
   at Avalonia.Collections.AvaloniaList`1.get_Item(Int32 index)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.TopLevel.HandleClosed()
   at Avalonia.Controls.WindowBase.HandleClosed()
   at Avalonia.Controls.Window.HandleClosed()
   at Avalonia.X11.X11Window.Cleanup(Boolean fromDestroyNotification)
   at Avalonia.X11.X11Window.Dispose()
   at Avalonia.Controls.Window.CloseInternal()
   at Avalonia.Controls.Window.HandleClosing(WindowCloseReason reason)
   at Avalonia.X11.X11Window.OnEvent(XEvent& ev)
   at Avalonia.X11.X11PlatformThreading.HandleX11(CancellationToken cancellationToken)
   at Avalonia.X11.X11PlatformThreading.RunLoop(CancellationToken cancellationToken)
   at Avalonia.Threading.DispatcherFrame.Run(IControlledDispatcherImpl impl)
   at Avalonia.Threading.Dispatcher.PushFrame(DispatcherFrame frame)
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime(AppBuilder builder, String[] args, ShutdownMode shutdownMode)
   at NexusMods.App.Startup.Main(IServiceProvider provider, String[] args) in /ssd-pool/projects/NexusMods/app/main/src/NexusMods.App/Startup.cs:line 19
   at NexusMods.App.StartupHandler.StartUiWindowAsync() in /ssd-pool/projects/NexusMods/app/main/src/NexusMods.App/StartupHandler.cs:line 52
   at NexusMods.SingleProcess.StartupDirector.Start(String[] args, Boolean debugMode, IAnsiConsole console)
```

## To Reproduce

This exception was thrown inside the `OnDetachedFromVisualTreeCore` method of a `Grid` that contained two or more children. The `Grid` was a used as an `ItemsPanelTemplate` in an `ItemsControl` that was bound to a `ReadOnlyObservableCollection<T>`.

## Environment

- OS: N/A
- Avalonia-Version: 11.0.7

## Additional context

Similar to #13497

### PR DESCRIPTION
**Pull request by TomEdwardsEnscape**
Create an "arena" at the start of a change to child controls, and copy the current state of the child collection into it before beginning iteration over that `Span`.

Fixes exceptions when children are removed, and fixes bugs when (pre-existing) children are not processed due to new children being inserted into the collection.

## Breaking changes
None

## Obsoletions / Deprecations
None

## Fixed issues
Fixes #13497
Fixes #12799
Fixes #14437

### CODE REVIEW

**Comment by MrJul** [comment]
I'm a bit worried about the potential performance impact here. This code path is hit for every element that's part of the attached tree - this can get large (take any treeview with a bunch of elements and watch allocations.

We've got three extra allocations for each call: one for the array, one for the action delegate and one for the closure. The action and closure allocations can be removed by having a state argument and a static method, but that's still leaves the array allocation and a full copy each time.

Plus the newly added/removed children aren't processed (which I admit is better than crashing, but might lead to other bugs later).

Were you able to identify which controls added or removed children while the enumeration was active? I think we should fix the root cause if possible.

**Comment by TomEdwardsEnscape** [comment]
I can reduce the allocations. Not sure how to avoid the array though.

> Were you able to identify which controls added or removed children while the enumeration was active? I think we should fix the root cause if possible.

There is no root cause, because user code could be doing anything in response to the methods that are being called. Perhaps a style will deactivate, causing an attached behaviour to turn off? We have no way of knowing what the side effects will be, so need to be prepared for anything.

> Plus the newly added/removed children aren't processed (which I admit is better than crashing, but might lead to other bugs later).

In all cases in this PR, that's OK. The code which mutates the child collection will (…or should…) already have executed the correct methods.

**Comment by workgroupengineering** [comment]
You can use src/Avalonia.Base/Utilities/SafeEnumerableList.cs or src/Avalonia.Base/Collections/Pooled/PooledList.cs

**Comment by TomEdwardsEnscape** [comment]
Unfortunately neither of these will help much in this case. `SafeEnumerableList` will generate a copy of the collection for each enumeration, which we don't want. `PooledList` is better as it will re-use arrays, but because the arrays in this case are being generated within a recursive call, a parent's array will remain in use until all of its children have been processed, and so we still have lots of allocations.

**Comment by maxkatz6** [comment]
Can we use CollectionMarshal.AsSpan here? Instead of making copy of the list, we can take inner array directly.
Something like this: https://github.com/AvaloniaUI/Avalonia/commit/b147f0d88329fdafcc8ecd12cab4cff9ccd14c6f

Essentially we take a Span on the inner array of the list.
If an item was added while iterating - it won't be included in our span (same behavior as currently).
If an item was removed while iterating - that index spot will [return null](https://github.com/dotnet/runtime/blob/f39dae85e28de85f87bb921a9e6a6581d531d293/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L1012). Luckily, we already have a `is StyledElement` condition there to filter nulls out.

Also:
1. Iterating over span is supposed to be faster than over list (no bounds checks).
2. I noticed that these methods use LogicalChildren property which creates this list on get-call. For these specific methods it feels unnecessary.
3. We can remove recursion completely by using PooledQueue. Not sure if we really need it though.

I haven't made any benchmarks yet (it's already late here), but in general it seems to work fine.
However, I am not sure if mine netstandard polyfill is bulletproof.

**Comment by maxkatz6** [comment]
cc @MrJul

**Comment by MrJul** [comment]
I like this from a performance standpoint, there should be absolutely no overhead (as you mentioned it should even be faster): no allocations, no copying.

I'm more worried about the polyfill, which is exploiting the internal layout of the type. Relying on such implementation details feels a bit wrong and dirty. That being said, it's only a polyfill for .NET Framework users, which doesn't evolve much, so it should be """safe""" (yes, triple quotes) enough.

I would use another class than `StrongBox<T>` to access the items though, for example a custom `class ListLayout<T> { public T[] Items; }`. This way we're relying only on `List<T>`'s layout not to change, instead of both `List<T>` + `StrongBox<T>`.

If this is implemented, some short code comment might be needed next to the children iteration to explain why we're using `CollectionsMarshal.AsSpan` here.

(Regarding `PooledQueue<T>`, I agree this probably isn't necessary - it's a larger change that should be measured first, out of scope of this PR imo.)

**Comment by TomEdwardsEnscape** [comment]
`AsSpan` won't work. Lists just don't work that way. [The method's documentation carries this warning:](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.asspan?view=net-7.0)

> Items should not be added or removed from the List<T> while the Span<T> is in use.

This is because:

- When adding, `List<T>` will only reallocate its internal array [if it reaches its current capacity](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs,201). When inserting in the middle of the list, [it will shunt items forward](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs,783).
- When removing, [it will shunt items backwards](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs,1010) so that the memory is contiguous.

In both cases we run into the exact same problems as looping over the collection directly.

**Comment by MrJul** [comment]
Yes, you're right, we'll potentially process the same items several times when removing. (Missing new ones due to an array reallocation following an insert isn't a new problem though, as it's not handled by copying the initial items either.)

**Comment by TomEdwardsEnscape** [comment]
I added the Arena class that I was talking about in an earlier comment, what are people's thoughts on that?

**Comment by MrJul** [comment]
> I added the Arena class that I was talking about in an earlier comment, what are people's thoughts on that?

While I like the principle, the fact that the used part of the arena is never released means we'll always allocate every N processed items.

Note that I'm nitpicking here! :) I'm aware that a few allocations here and there don't mean much in the grand scheme of things (there are way worse allocations happening currently). It can be mitigated by using a slightly larger chunk size.

(Whatever we choose should be an implementation detail, so if it's the Arena it should be `internal`, but I'll review it properly if we go this route.)

Waiting for @maxkatz6 thoughts on this.

**Comment by TomEdwardsEnscape** [comment]
@maxkatz6 bump!

**Comment by Gillibald** [comment]
Can we investigate why logical children change during tree detachment? This is some invalid behavior. Your changes just work around it when we should fix the real cause.

**Comment by TomEdwardsEnscape** [comment]
It shouldn't be considered invalid behaviour, because detaching a control can have user-defined side effects. [See my earlier comment:](#issuecomment-1796052437)

> There is no root cause, because user code could be doing anything in response to the methods that are being called. Perhaps a style will deactivate, causing an attached behaviour to turn off? We have no way of knowing what the side effects will be, so need to be prepared for anything.

**Comment by Gillibald** [comment]
Suppose the user code modifies logical children during enumeration that is the user's fault. We could guard enumeration in some way. WPF usually sets a flag so subsequent callers can queue an action that gets executed in the next dispatcher frame.

**Comment by grokys** [comment]
Note that WPF doesn't handle modification of the logical tree during iteration either:

https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Generated/FrameworkElement.cs#L198

**Comment by TomEdwardsEnscape** [comment]
> I'm not suggesting to use `SafeEnumerableList` here, but to use the same technique of copy-on-write.

To be performant, this would need to be built into `AvaloniaList`. It could return a span that points to its current internal array, then right before any change is made copy its current state (perhaps via the `Arena` type in this PR) and redirected the span to that memory. The hard part is knowing when to stop doing that without doing stuff like implementing `IDisposable` or keeping a list of subscribers, both of which require allocations for each iteration.

Another idea is to create an "INotifyCollectionChanging" interface for `AvaloniaList`, and then have `StyledElement` copy the children into the arena in response to the event. But this involves creating lots of event subscriptions at each level of the tree and each one is another heap allocation.

TLDR: it's hard to avoid allocations!

**Comment by TomEdwardsEnscape** [comment]
I added an exception like WPF's, which is thrown if the logical children collection is changed during internal enumeration. [Here is the new call stack](https://gist.github.com/TomEdwardsEnscape/5487234793b086912c2864718c4417ed) for the repro steps from #13497.

What happens:

1. The `TabControl` is removed from the logical tree
2. The `TabControl` can no longer see the styles inherited from its parent
3. The active `TabItem` loses the `ContentTemplate` previously provided by one of the inherited styles
4. The `TabItem`'s `ContentPresenter` [removes its "effective" logical child](https://github.com/AvaloniaUI/Avalonia/blob/18d02f1aa005b87ff85c995751b73e02f7e38567/src/Avalonia.Controls/Presenters/ContentPresenter.cs#L704), which surprisingly is:
    * The `Border` generated by the `ContentTemplate`, and not the content being presented
    * A child of the parent `TabControl`, not the `TabItem`
6. Because the `TabControl`'s children have been modified by its `TabItem`, the exception is thrown

In WPF, the only logical child of the `TabControl` is the `TabItem`. In Avalonia we have this:

image

[This ties into some earlier observations I made on Avalonia's (mis)use of the logical tree.](https://github.com/AvaloniaUI/Avalonia/issues/13230#issuecomment-1774050018)

**Comment by erri120** [comment]
@TomEdwardsEnscape are you sure this also fixes #14437? That issue is for the _visual_ instead of the _logical_ tree.

**Comment by TomEdwardsEnscape** [comment]
Hadn't spotted that, thanks for the tip! The PR now handles visual child enumeration too.

File: src/Avalonia.Base/Collections/Arena.cs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+namespace Avalonia.Collections;
+
+/// <summary>
+/// Reduces memory allocations by handing out <see cref="Span{T}"/> views of larger array chunks.
+/// </summary>
+/// <remarks>
+/// Intended for scenarios in which many short-lived collections are required simultaneously.
+/// </remarks>
+internal class Arena<T>
```
Comment (kekekeks) [inline src/Avalonia.Base/Collections/Arena.cs:14]: Am I missing something or is this just a worse version of an array pool? 
It doesn't really support reentrancy since it tries to use a single array per thread, allocates unpooled memory  and uses GCHandle for a reason that escapes me

Comment (kekekeks) [reply src/Avalonia.Base/Collections/Arena.cs:14]: It also doesn't seem to clear the array once it's freed, which can lead to memory leaks

Comment (kekekeks) [reply src/Avalonia.Base/Collections/Arena.cs:14]: 1) I still don't see where the array gets cleared. The "current" array is GC-rooted, so it will cause a temporary memory leak (which can and will GC-root huge trees of objects in user apps)
2) Previously we didn't allocate at all. Why can't we use an ArrayPool? Why do we need to be constantly allocating new arrays?

**Comment by MrJul** [comment]
Due to the disagreements regarding the implementation, I'm going to close this.
We should implement a proper copy-on-write `IAvaloniaList` instead: see #17236
